### PR TITLE
processor/k8sattributesprocessor: map containers by ID

### PR DESCRIPTION
**Pull request by gbbr**
This change allows the user (or SDK) sent [`container.id`](https://github.com/open-telemetry/opentelemetry-go/blob/sdk/v1.14.0/sdk/resource/config.go#L188-L201) to be picked up by the processor and used to identify container attributes. As a result, it is now possible to obtain the `k8s.container.name` attribute (along with the `container.image.{name,tag}` pair) more easily and in line with the SDK API.

Closes #19468

### CODE REVIEW

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 1]
```diff
@@ -374,12 +374,26 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
...
+		return containers
+	}
+	ifRule := func(rule bool, v string) string {
+		if rule {
+			return v
+		}
+		return ""
+	}
+	switch {
+	case c.Rules.ContainerImageName,
```
Comment (gbbr) [inline processor/k8sattributesprocessor/internal/kube/client.go:392]: Side-note (@dmitryax): I think complexity of this code (and overall) could be greatly reduced if we would _always_ set all fields of these structs, and only check the rules when setting the actual attributes. It seems like concerns would be better separated that way and the complexity here reduced. WDYT? I suspect it wasn't memory that was a concern here. If it was, that can still be addressed (better and) differently.

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 2]
```diff
@@ -374,12 +374,26 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
...
+	if !needContainerAttributes(c.Rules) {
+		return containers
+	}
+	ifRule := func(rule bool, v string) string {
+		if rule {
+			return v
+		}
+		return ""
+	}
+	switch {
```
Comment (gbbr) [inline processor/k8sattributesprocessor/internal/kube/client.go:391]: Side note (cc @dmitryax): I would like to propose to remove all references to the rules from this package. That would then greatly simplify the complexity here. We can simply check the rules in the processor and avoid setting attributes if rules aren't set. This centralises and concentrates that logic in one places, making it much easier to reason about.

If performance was the concern and reason for the current architecture, then that can easily be addressed in another way.

If you like this idea, happy to propose & work on as a separate issue/PR.

Comment (dmitryax) [reply processor/k8sattributesprocessor/internal/kube/client.go:391]: I'm always up for simplifying code! The only requirement is not to introduce any additional k8s API calls.  Please feel free to submit a separate PR

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 3]
```diff
@@ -651,5 +663,12 @@ func (c *WatchClient) extractNamespaceLabelsAnnotations() bool {
...
 func needContainerAttributes(rules ExtractionRules) bool {
-	return rules.ContainerImageName || rules.ContainerImageTag || rules.ContainerID
+	switch {
+	case rules.ContainerImageName,
+		rules.ContainerName,
+		rules.ContainerImageTag,
+		rules.ContainerID:
+		return true
+	}
+	return false
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:673]: Not sure I agree with this style. The former version seems more readable to me :)

File: processor/k8sattributesprocessor/options.go [hunk 4]
```diff
@@ -99,6 +99,8 @@ func withExtractMetadata(fields ...string) option {
 				p.rules.StatefulSetName = true
 			case conventions.AttributeK8SStatefulSetUID:
 				p.rules.StatefulSetUID = true
+			case conventions.AttributeK8SContainerName:
+				p.rules.ContainerName = true
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/options.go:103]: Please update readme and config.go. I believe we list all supported metadata options there

File: processor/k8sattributesprocessor/processor.go [hunk 5]
```diff
@@ -151,46 +151,64 @@ func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pco
...
-			attrs.PutStr(conventions.AttributeContainerImageTag, containerSpec.ImageTag)
+	case containerID != "":
+		containerSpec, ok = pod.Containers.ByID[containerID]
+		if !ok {
+			return
 		}
+	default:
+		return
 	}
-
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:172]: Are all the changes below needed for this change or just refactoring?

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 1 above)
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:392]: nit: I don't think this switch + one case approach is used across the repo. Let's not introduce new idioms

**Review by dmitryax — COMMENTED** [review]
Can you please keep the changeset minimal to only what's required for the new functionality. If you want apply refactoring, let's do that separately.

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 2 above)
Comment (gbbr) [reply processor/k8sattributesprocessor/internal/kube/client.go]: Great. Will do. Thanks.

File: processor/k8sattributesprocessor/processor.go (hunk 5 above)
Comment (gbbr) [reply processor/k8sattributesprocessor/processor.go:172]: It's basically a refactoring. I tried to avoid duplicating that code even more. I will revert for the sake of keeping the PR focused on the change. I presume that's what you're trying to say.

**Review by gbbr — COMMENTED** [review]
Thanks for the review. I think I addressed all of your comments. PTAL.

File: processor/k8sattributesprocessor/processor.go [hunk 6]
```diff
@@ -151,46 +151,60 @@ func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pco
...
+	if _, ok := attrs.Get(conventions.AttributeK8SContainerName); !ok && containerSpec.Name != "" {
+		attrs.PutStr(conventions.AttributeK8SContainerName, containerSpec.Name)
 	}
-
-	if containerSpec.ImageName != "" {
-		if _, found := attrs.Get(conventions.AttributeContainerImageName); !found {
-			attrs.PutStr(conventions.AttributeContainerImageName, containerSpec.ImageName)
-		}
+	if _, ok := attrs.Get(conventions.AttributeContainerImageName); !ok && containerSpec.ImageName != "" {
+		attrs.PutStr(conventions.AttributeContainerImageName, containerSpec.ImageName)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:177]: This adds a redundant call to `attrs.Get` which I don't think will be optimized by the compiler. Should be done in another refactoring PR with a benchmark, ideally.

File: processor/k8sattributesprocessor/processor.go [hunk 7]
```diff
@@ -151,46 +151,60 @@ func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pco
...
 	}
-	if containerSpec.ImageTag != "" {
-		if _, found := attrs.Get(conventions.AttributeContainerImageTag); !found {
-			attrs.PutStr(conventions.AttributeContainerImageTag, containerSpec.ImageTag)
-		}
+	if _, ok := attrs.Get(conventions.AttributeContainerImageTag); !ok && containerSpec.ImageTag != "" {
+		attrs.PutStr(conventions.AttributeContainerImageTag, containerSpec.ImageTag)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:180]: Same here

File: processor/k8sattributesprocessor/processor.go (hunk 5 above)
Comment (dmitryax) [reply processor/k8sattributesprocessor/processor.go:172]: Reusing code is fine. Just wanted to make sure that the changeset doesn't have unrelated refactoring like the ones above

File: processor/k8sattributesprocessor/processor.go [hunk 8]
```diff
@@ -151,46 +151,60 @@ func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pco
...
-	runIDAttr, ok := attrs.Get(conventions.AttributeK8SContainerRestartCount)
-	if ok {
-		containerRunID, err := intFromAttribute(runIDAttr)
-		if err != nil {
-			kp.logger.Debug(err.Error())
-		} else {
-			runID = containerRunID
-		}
+	if _, ok := attrs.Get(conventions.AttributeContainerID); !ok && containerSpec.ID != "" {
+		attrs.PutStr(conventions.AttributeContainerID, containerSpec.ID)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:183]: `ID` cannot be part of `containerSpec`. Every container status has its own ID. Looks like here we will always be setting `conventions.AttributeContainerID` attribute to the latest container status ignoring provided `conventions.AttributeK8SContainerRestartCount`

Comment (gbbr) [reply processor/k8sattributesprocessor/processor.go]: I see what you mean. Let me revert this change back to what it was.

**Review by gbbr — COMMENTED** [review]
Addressed.

File: processor/k8sattributesprocessor/processor.go [hunk 9]
```diff
@@ -151,25 +151,41 @@ func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pco
...
-		return
+	if containerSpec.Name != "" {
+		if _, ok := attrs.Get(conventions.AttributeK8SContainerName); !ok {
+			attrs.PutStr(conventions.AttributeK8SContainerName, containerSpec.Name)
+		}
 	}
-
 	if containerSpec.ImageName != "" {
-		if _, found := attrs.Get(conventions.AttributeContainerImageName); !found {
+		if _, ok := attrs.Get(conventions.AttributeContainerImageName); !ok {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:179]: nit: please avoid unrelated changes in the diff

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 10]
```diff
@@ -374,12 +374,23 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
...
+	if !needContainerAttributes(c.Rules) {
+		return containers
+	}
+	newContainer := func(name string) *Container {
+		if c.Rules.ContainerName {
+			return &Container{Name: name}
+		}
+		return &Container{}
+	}
+	if c.Rules.ContainerImageName || c.Rules.ContainerImageTag || c.Rules.ContainerName {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:391]: Do we really need to go over this loop if `c.Rules.ContainerName`? I think we can avoid that and assign this in the second loop which we have to go through anyway

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 11]
```diff
@@ -374,12 +374,23 @@ func (c *WatchClient) extractPodAttributes(pod *api_v1.Pod) map[string]string {
...
 		for _, spec := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
-			container := &Container{}
+			container := newContainer(spec.Name)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:393]: In that case this change isn't needed

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 12]
```diff
@@ -391,29 +402,26 @@ func (c *WatchClient) extractPodContainersAttributes(pod *api_v1.Pod) map[string
...
-				container = &Container{}
-				containers[apiStatus.Name] = container
-			}
+	for _, apiStatus := range append(pod.Status.ContainerStatuses, pod.Status.InitContainerStatuses...) {
+		container, ok := containers.ByName[apiStatus.Name]
+		if !ok {
+			container = newContainer(apiStatus.Name)
+			containers.ByName[apiStatus.Name] = container
+		}
+		containerID := apiStatus.ContainerID
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:414]: And this becomes 
```suggestion
		container, ok := containers.ByName[apiStatus.Name]
		if !ok {
			container = &Container{}
			containers.ByName[apiStatus.Name] = container
		}
		if c.Rules.ContainerName {
			container.Name = name
		}
		
		containerID := apiStatus.ContainerID
```

so we don't need `newContainer` func anymore

**Comment by dmitryax** [comment]
Please take a look at the failing linter

File: processor/k8sattributesprocessor/processor.go (hunk 9 above)
Comment (gbbr) [reply processor/k8sattributesprocessor/processor.go]: Ah, this is a habit of mine that doesn't apply on this repository... Sorry, fixed.

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 12 above)
Comment (gbbr) [reply processor/k8sattributesprocessor/internal/kube/client.go]: True. This works. It's a bit confusing. That's why I [suggested](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20340#discussion_r1149112774) to remove all the conditions related to rules from this package. I'll try to address that separately.

Thanks for your patience.

**Review by dmitryax — APPROVED** [review]

**Comment by gbbr** [comment]
Thanks Dmitrii!
